library/core/src/main/java/com/google/android/exoplayer2/text/ttml/TextEmphasis.java

```diff
@@ -0,0 +1,189 @@
+   */
+  static final int POSITION_OUTSIDE = 1 << 8;
+
+  @Documented
+  @Retention(SOURCE)
+  @IntDef({TextAnnotation.POSITION_UNKNOWN,
+      TextAnnotation.POSITION_BEFORE,
+      TextAnnotation.POSITION_AFTER,
+      // Extending the definition in TextAnnotation.Position for intermediate values
+      POSITION_OUTSIDE
+  })
+  @interface Position {}
+
+  /**
+   * The position of the text emphasis relative to the base text.
+   */
+  @Position final int position;
+
+  private static Set markValues = ImmutableSet.of(
+      TtmlNode.TEXT_EMPHASIS_AUTO,
+      TtmlNode.TEXT_EMPHASIS_MARK_DOT,
+      TtmlNode.TEXT_EMPHASIS_MARK_SESAME,
+      TtmlNode.TEXT_EMPHASIS_MARK_CIRCLE
+      );
+
+  private static Set markStyles = ImmutableSet.of(
+      TtmlNode.TEXT_EMPHASIS_MARK_FILLED,
+      TtmlNode.TEXT_EMPHASIS_MARK_OPEN
+  );
+
+  private static Set positionValues = ImmutableSet.of(
+      TtmlNode.ANNOTATION_POSITION_AFTER,
+      TtmlNode.ANNOTATION_POSITION_BEFORE,
+      TtmlNode.ANNOTATION_POSITION_OUTSIDE
+  );
+
+  private TextEmphasis(@Mark int mark, @TextAnnotation.Position int position) {
+    this.mark = mark;
+    this.position = position;
+  }
+
+  @Override
+  public String toString() {
+    return "TextEmphasis{" +
+        "position=" + position +
+        ", mark=" + mark +
+        '}';
+  }
+
+  public static TextEmphasis createTextEmphasis(@Nullable String value) {
+    if (value == null) {
+      return null;
+    }
+
+    String parsingValue = value.toLowerCase().trim();
+    if ("".equals(parsingValue)) {
+      return null;
+    }
+
+    Set<String> nodes = Sets.newHashSet(parsingValue.split("\\s+"));
+
+    if (nodes.size() == 0 || TtmlNode.TEXT_EMPHASIS_NONE.equals(nodes.iterator().next())) {
+      return null;
+    }
+    return parseNodes(nodes);
+  }
+
+  private static @Nullable TextEmphasis parseNodes(Set<String> nodes) {
+    Set styleSet = Sets.intersection(markStyles, nodes).immutableCopy();
+    Set markSet = Sets.intersection(markValues, nodes).immutableCopy();
+    Set positionSet = Sets.intersection(positionValues, nodes).immutableCopy();
+
+    @Mark int mark = 0;
+    if (styleSet.size() == 1) {
+      mark |= TtmlNode.TEXT_EMPHASIS_MARK_OPEN.equals(styleSet.iterator().next())
+          ? TextEmphasisSpan.MARK_FLAG_OPEN
+          : TextEmphasisSpan.MARK_FLAG_FILLED;
+    }
+    if (markSet.size() == 1) {
+      switch ((String) markSet.iterator().next()) {
+        case TtmlNode.TEXT_EMPHASIS_AUTO:
+          mark |= MARK_AUTO;
+          break;
+        case TtmlNode.TEXT_EMPHASIS_MARK_DOT:
+          mark |= TextEmphasisSpan.MARK_FLAG_DOT;
+          break;
+        case TtmlNode.TEXT_EMPHASIS_MARK_SESAME:
+          mark |= TextEmphasisSpan.MARK_FLAG_SESAME;
+          break;
+        case TtmlNode.TEXT_EMPHASIS_MARK_CIRCLE:
+        default:
+          mark |= TextEmphasisSpan.MARK_FLAG_CIRCLE;
+      }
+    } else {
+      mark |= TextEmphasisSpan.MARK_FLAG_CIRCLE;
```

Comment:
Good catch! I think depends on the number of valid arguments. i.e. "filled", vs. "invalid". Both would result in an empty `markSet`. The former should be set to `circle`, while the latter should be set to `auto`

> If only filled or open is specified, then it is equivalent to filled circle and open circle, respectively.

Reply:
This gets a little tricky. Consider the following if the user makes a typo:
`filed circle` -> equivalent to "auto" or "circle"?
`open sesamee` -> equivalent to "auto" or "open"?
Based on the wording of the spec, one could conclude that this is an unrecognized emphasis style value, and should be treated as "auto". What do you think?

Reply:
On further thought, there is no way to know whether the invalid attribute was intended to be a style value or position value or color (not supported) value. This is my revised position on this:
"filed circle" -> equivalent to "circle"
"open sesamee" -> equivalent to "open"
wdyt?

Reply:
I don't think 'filed' can match a color, because they [either start with `#` or `rgb`](https://www.w3.org/TR/2018/REC-ttml2-20181108/#style-value-color) or are one of the [19 predefined named colors](https://www.w3.org/TR/2018/REC-ttml2-20181108/#style-value-named-color).

I assume the `<quoted-string>` part of `emphasis-style` would have **extra** quotes around it (as well as not being supported by your impl here)? i.e. something like `'foo' blue` - in which case `filed` can't be mistaken for a quoted string either.

So I think a rigid interpretation of the spec probably rejects `filed blue` completely as malformed (i.e. doesn't match anything). Not sure what the TTML spec expects in that case, imo it seems reasonable to reject the entire node and behave as if it never existed (probably logging a warning).

Implementing that with the existing set-based logic might be a bit tricky to do though (and arguably user-hostile - it's a tricky line between being over-precise and over-tolerant) - so it's also probably fine to treat it like `circle` as you propose. It doesn't seem worth spending ages handling a misspelt edge-case...

Reply:
I understand what you're saying that user-intent can be inferred even in misspelled attributes. What I am suggesting is that a typo on a non-quoted string will still be ambiguous. i.e.`b` could be intended for color, position or styling. 
In the example you gave, i.e. `filed blue`, my interpretation of the spec is that `auto` styling should apply. If we had an implementation that supported the color attribute, then the equivalent of `auto blue` should apply.

Reply:
> In the example you gave, i.e. `filed blue`, my interpretation of the spec is that `auto` styling should apply. If we had an implementation that supported the color attribute, then the equivalent of `auto blue` should apply.

Interesting - are you approaching this from the perspective of "the code should be written to allow the spec to evolve later"? Because with the **current** spec I don't see how you go from `filed blue` to `auto blue`. `filed` isn't an `emphasis-style`, an `emphasis-color` or an `emphasis-position` - right? That's why it seems to me the whole line should strictly be considered invalid.

If the code needs to assume there might be un-quoted `emphasis-style` names that were added to the spec since the code was written (or predefined color names I guess, bleh) - then I get your point about ambiguity. I think that's a reasonable & realistic stance, because of course specs evolve. And resolving unknown things to the `auto` behaviour seems like a pretty safe default, so I agree we should go with that.

Do you think it's worth adding some additional cues to `text_emphasis.xml` with some 'invalid' lines like this to show we thought about how they should resolve?

Reply:
I wasn't approaching it from the perspective that spec might evolve later, at least not consciously. Let me see if I can explain it better.

> <text-emphasis>   : <emphasis-style> \|\| <emphasis-color> \|\| <emphasis-position>
--

The question I am debating is whether we consider the validity of each attribute (style, color, position) value independently? For example, "filed blue' has a valid color value, but an unspecified or invalid style or position value. Do we 1) disregard the entire description and either revert to `auto` or `none` or 2) apply the valid values but revert to defaults for the invalid or unspecified values?

For approach 1) we would be looking for values outside the set of valid values. 
`filed blue` would resolve to `auto` | `none`
`filled dot blue after foo` would resolve to `auto` | `none`
This is a more rigid interpretation of the spec, and less future-friendly if the specs evolve, requires full implementation of the spec (parsing color and quoted string styles) but more deterministic in that any invalid descriptions would always resolve to the same result.

For approach 2) we would be looking for values in the set of valid values and applying the ones we detect, and discarding the rest.
`filed blue` would resolve to `auto blue`
`filled dot blue after foo` would resolve to `filled dot blue after`
This is a looser interpretation of the spec, and is more future-friendly if the specs evolve, but less deterministic as partial features may be applied (i.e. color only but not style or position).

The current implementation takes approach 2), but after more consideration, I believe the correct interpretation of the spec should be 1), which I believe is your position already.

For reference, the IMSC JS parser http://sandflow.com/imsc1_1/index.html takes approach 1).
All three browsers (Firefox, Chrome, Safari) takes approach 1) based on the CSS attributes for `text-emphasis`.
All of the above implementations fallback to `none` if the description contains invalid values.

I'd love to hear your thoughts before I make any changes.

Reply:
Thanks for explaining further - I understand now.

Thanks for pointing out that my 'strict' proposal implicitly requires a **full** spec implementation, including quoted strings and colors - I hadn't fully appreciated that (though seems obvious in retrospect). Or at least the parser has to be aware of the full spec, even if it then throws away bits it can't implement and reverts them to `auto`.

The fact that the existing IMSC parser, and browsers for the CSS property, all take the strict approach does suggest we should ideally do the same. On a pragmatic level though, it seems pretty unlikely that approach 2 will be truly problematic...

I think there are three options here:
a. Merge this PR as-is with approach 2.
b. Change this PR to approach 1 and add full parse support for colors & quoted strings (even if they're not supported by TextEmphasisSpan or the HTML/CSS rendering yet).
c. Do (a) and then do (b) in a follow-up PR.

I'll leave it up to you - I'm happy to take it as-is with no plan to make it stricter in future, but also happy to work with you on either (b) or (c). Let me know what you prefer!

Reply:
I propose we do a) as it addresses the typical use cases for Japanese subtitles. Additionally, I'll add comments to explain the approach as well as the edge case examples in the tests as you suggested.

This doesn't preclude b) from happening in the future, but we can just make it low priority for now.

Reply:
Yep sounds good - I'll work on getting this merged. Thanks!